Add docs related to integration of Hashicrop Vault extension (#8432)
wso2/docs-apim: en/docs/install-and-setup/setup/security/logins-and-passwords/harshicrop-vault-extension.md

   
1. Using Root Token authentication
2. Using App-Role authentication

## Setting up using Root Token authentication

This method uses a static root token to authenticate with HashiCorp Vault, providing direct and full access to Vault's secrets.

### Step 1 - Setup HashiCorp Vault

1. Start the HashiCorp Vault server and create a new **KV engine**. Enter a path name when creating the kv engine (e.g., `wso2apim`). The following commands can be used to add secrets with the HashiCorp Vault.

    ```shell
    # Create a new kv engine
    vault secrets enable -path=wso2apim -version=2 kv

    # Add new secret
    vault kv put wso2apim/admin_password value=admin

    # Get a secret (To check)
    vault kv get -field=value wso2apim/admin_password
    ```

### Step 2 - Configure HashiCorp Vault extension

1. Clone [HashiCorp Vault extension](https://github.com/wso2-extensions/carbon-securevault-hashicorp/tree/master) repo.

2. Build the HashiCorp Vault Integration OSGI bundle using `mvn clean install` command and copy

>    the `target/org.wso2.carbon.securevault.hashicorp-1.0.jar` file to the `<API-M_HOME>/repository/components/dropin/`

Author: fixed in https://github.com/wso2/docs-apim/pull/9641

# Integrate with HashiCorp Vault

Using [HashiCorp Vault extension](https://github.com/wso2-extensions/carbon-securevault-hashicorp/tree/master), you can set up HashiCorp Vault to store passwords that are mapped to aliases instead of the actual passwords.

## Setting up using root token authentication

> 

Reviewer: If possible we can add a sentence explaining what this process is

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> ### Step 1: Setup HashiCorp Vault

Reviewer: ```suggestion
### Step 1 - Setup HashiCorp Vault
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Start the HashiCorp Vault server and create a new **kv engine**. Enter a path name when creating the kv engine (e.g., `wso2apim`). The following commands can be used to add secrets with the HashiCorp Vault.

Reviewer: ```suggestion
1. Start the HashiCorp Vault server and create a new **KV engine**. 
2. Enter a path name when creating the KV engine (e.g., `wso2apim`). The following commands can be used to add secrets with the HashiCorp Vault.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```bash
    # Create a new kv engine
    vault secrets enable -path=wso2apim -version=2 kv

    # Add new secret
    vault kv put wso2apim/admin_password value=admin

    # Get a secret (To check)
    vault kv get -field=value wso2apim/admin_password
    ```

> ### Step 2: Configure HashiCorp Vault extension

Reviewer: ```suggestion
### Step 2 - Configure HashiCorp Vault extension
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Clone [HashiCorp Vault extension](https://github.com/wso2-extensions/carbon-securevault-hashicorp/tree/master) repo

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 2. Build the HashiCorp Vault Integration OSGI bundle using `mvn clean install` and copy

Reviewer: ```suggestion
2. Build the HashiCorp Vault Integration OSGI bundle using `mvn clean install` command and copy
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

>    the `target/org.wso2.carbon.securevault.hashicorp-1.0.jar` file to the `<APIM_HOME>/repository/components/dropin/`

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

   directory.

3. Add **HashiCorp Vault Java Driver** (e.g., `vault-java-driver-5.1.0.jar`) to the

>    `<APIM_HOME>/repository/components/lib/` directory.

Reviewer: ```suggestion
   `<API-M_HOME>/repository/components/lib/` directory.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```properties
    keystore.identity.location=repository/resources/security/wso2carbon.jks
    keystore.identity.type=JKS
    keystore.identity.store.password=identity.store.password
    keystore.identity.store.secretProvider=org.wso2.carbon.securevault.DefaultSecretCallbackHandler
    keystore.identity.key.password=identity.key.password
    keystore.identity.key.secretProvider=org.wso2.carbon.securevault.DefaultSecretCallbackHandler
    carbon.secretProvider=org.wso2.securevault.secret.handler.SecretManagerSecretCallbackHandler

    secVault.enabled=true
    secretRepositories=vault
    secretRepositories.vault.provider=org.wso2.carbon.securevault.hashicorp.repository.HashiCorpSecretRepositoryProvider
    secretRepositories.vault.properties.address=https://127.0.0.1:8200
    secretRepositories.vault.properties.namespace=ns1
    secretRepositories.vault.properties.enginePath=wso2apim
    secretRepositories.vault.properties.engineVersion=2
    secretRepositories.vault.properties.authType=ROOT_TOKEN
    ```

    !!! note
        In production, you should always use the vault address with TLS enabled.

> 5. Add the following lines to the `<APIM_HOME>/repository/conf/log4j2.properties` file:

Reviewer: ```suggestion
5. Add the following lines to the `<API-M_HOME>/repository/conf/log4j2.properties` file:
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```properties
    logger.org-wso2-carbon-securevault-hashicorp.name=org.wso2.carbon.securevault.hashicorp
    logger.org-wso2-carbon-securevault-hashicorp.level=INFO
    logger.org-wso2-carbon-securevault-hashicorp.additivity=false
    logger.org-wso2-carbon-securevault-hashicorp.appenderRefCARBON_CONSOLE. ref = CARBON_CONSOLE
    ```

> 6. Then append `org-wso2-carbon-securevault-hashicorp` to the `loggers` list in the same file as follows:

Reviewer: ```suggestion
6. Then append `org-wso2-carbon-securevault-hashicorp` to the `loggers` list in the same file as follows.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```properties
    loggers = AUDIT_LOG, trace-messages, ..., org-wso2-carbon-securevault-hashicorp
    ```

> ### Step 3: Update passwords with their aliases

Reviewer: ```suggestion
### Step 3 - Update passwords with their aliases
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Open the `deployment.toml` file in the `<APIM_HOME>/repository/conf/` directory and add

Reviewer: ```suggestion
1. Open the `deployment.toml` file in the `<API-M_HOME>/repository/conf/` directory and add
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

   the `[secrets]` configuration section **at the bottom of the file** as shown below.
   Give an alias for the passwords and put the value as blank (`""`).

    ```toml
    [secrets]
    admin_password = ""
    ```

2. Add the encrypted password alias to the relevant sections in the `deployment.toml`
   file by using a placeholder: `$secret{alias}`. For example:

    ```toml
    [super_admin]
    username="admin"
    password="$secret{admin_password}"
    ```

> ### Step 4: Start the server

Reviewer: ```suggestion
### Step 4 - Start the server
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Provide the `VAULT_TOKEN` to the prompted message in the console or by creating a new file in the `<APIM_HOME>` 

Reviewer: ```suggestion
1. Provide the `VAULT_TOKEN` to the prompted message in the console or by creating a new file in the `<API-M_HOME>` 
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

   directory. The file should be named according to your Operating System.

    === "Linux"
        ```

>         The file name should be `hashicorpRootToken-tmp`.

Reviewer: Let's change the formatting accordingly as this is a statement, not a command. Same goes for the Windows tab.

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

        ```
      
    === "Windows"
        ```
        The file name should be `hashicorpRootToken-tmp.txt`.
        ```

    !!! note

>         When you add "tmp" to the file name, note that this will automatically get deleted from the file system after the server starts. Alternatively, if you want to retain the password file after the server starts, the file should be named as follows:

Reviewer: ```suggestion
        When you add `tmp` to the file name, note that this will automatically get deleted from the file system after the server starts. Alternatively, if you want to retain the password file after the server starts, the file should be named as follows:
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

        === "Linux"
            ```

>             The file name should be `hashicorpRootToken-persist`.

Reviewer: Let's change the formatting as suitable for a statement.

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

            ```
        
        === "Windows"
            ```
            The file name should be `hashicorpRootToken-persist.txt`.
            ```

> 2. Start the WSO2 API Manager Server and enter the keystore password at startup when prompted:

Reviewer: ```suggestion
2. Start the WSO2 API Manager Server and enter the keystore password at startup when prompted.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

>     ```bash

Reviewer: ```suggestion
    ```shell
```

Reviewer: Try using shell for formatting

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    [Enter KeyStore and Private Key Password :] wso2carbon
    ```

> ---

Reviewer: ```suggestion

```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> ## Setting up using AppRole authentication

Reviewer: Let's add a small intro in a sentence similar to the previous section of steps.

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> ### Step 1: Setup HashiCorp Vault

Reviewer: ```suggestion
### Step 1 - Setup HashiCorp Vault
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Start the HashiCorp Vault server and set the environment variables:

Reviewer: ```suggestion
1. Start the HashiCorp Vault server and set the environment variables.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```bash
    export VAULT_ADDR='http://127.0.0.1:8200'
    export VAULT_TOKEN='<root token>'
    ```

> 2. Include the policy in the `kv-read-write.hcl` file as below:

Reviewer: ```suggestion
2. Include the policy in the `kv-read-write.hcl` file as below.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```hcl
    path "wso2apim/data/*" {
        capabilities = ["create", "read", "update", "delete", "list"]
    }
    ```

> 3. Upload the kv policy as below:

Reviewer: ```suggestion
3. Upload the kv policy using the command below:
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

>     ```bash

Reviewer: ```suggestion
    ```shell
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    vault policy write kv-read-write kv-read-write.hcl
    ```

> 4. The following commands can be used to create the AppRole using the created policy:

Reviewer: ```suggestion
4. The following commands can be used to create the AppRole using the created policy.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```bash
    vault auth enable approle
    ```
    ```bash   
    vault write auth/approle/role/my-role \
        token_policies="kv-read-write" \
        token_type="service" \
        token_ttl="24h" \
        token_max_ttl="72h"
    ```

> 5. Get the role-id as below:

Reviewer: ```suggestion
5. Get the role-id using the command below.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```bash
    vault read auth/approle/role/my-role/role-id
    ```

> 6. Get the secret-id as below:

Reviewer: ```suggestion
6. Get the secret-id using the command below:
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```bash
    vault write -f auth/approle/role/my-role/secret-id
    ```

7. Create the kv engine with a path name (e.g., `wso2apim`).

>     The following commands can be used to add secrets with the HashiCorp Vault:

Reviewer: ```suggestion
7. Create the kv engine with a path name (e.g., `wso2apim`). The following commands can be used to add secrets with the HashiCorp Vault:
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```bash
    # Create a new kv engine
    vault secrets enable -path=wso2apim -version=2 kv

    # Add new secret
    vault kv put wso2apim/admin_password value=admin

    # Get a secret (To check)
    vault kv get -field=value wso2apim/admin_password
    ```

> ### Step 2: Configure HashiCorp Vault extension

Reviewer: ```suggestion
### Step 2 - Configure HashiCorp Vault extension
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Clone [HashiCorp Vault extension](https://github.com/wso2-extensions/carbon-securevault-hashicorp/tree/master) repo

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 2. Build the HashiCorp Vault Integration OSGI bundle using `mvn clean install` and copy the `target/org.wso2.carbon.securevault.hashicorp-1.0.jar` file to the `<APIM_HOME>/repository/components/dropin/`

Reviewer: ```suggestion
2. Build the HashiCorp Vault Integration OSGI bundle using `mvn clean install` command and copy the `target/org.wso2.carbon.securevault.hashicorp-1.0.jar` file to the `<API-M_HOME>/repository/components/dropin/`
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

   directory.

3. Add **HashiCorp Vault Java Driver** (e.g., `vault-java-driver-5.1.0.jar`) to the

>    `<APIM_HOME>/repository/components/lib/` directory.

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 4. Create the `/repository/conf/security/secret-conf.properties` file and set the following configurations:

Reviewer: ```suggestion
4. Create the `/repository/conf/security/secret-conf.properties` file and set the following configurations.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```properties
    keystore.identity.location=repository/resources/security/wso2carbon.jks
    keystore.identity.type=JKS
    keystore.identity.store.password=identity.store.password
    keystore.identity.store.secretProvider=org.wso2.carbon.securevault.DefaultSecretCallbackHandler
    keystore.identity.key.password=identity.key.password
    keystore.identity.key.secretProvider=org.wso2.carbon.securevault.DefaultSecretCallbackHandler
    carbon.secretProvider=org.wso2.securevault.secret.handler.SecretManagerSecretCallbackHandler

    secVault.enabled=true
    secretRepositories=vault
    secretRepositories.vault.provider=org.wso2.carbon.securevault.hashicorp.repository.HashiCorpSecretRepositoryProvider
    secretRepositories.vault.properties.address=http://127.0.0.1:8200
    secretRepositories.vault.properties.namespace=ns1
    secretRepositories.vault.properties.enginePath=wso2apim
    secretRepositories.vault.properties.engineVersion=2
    secretRepositories.vault.properties.authType=APP_ROLE
    secretRepositories.vault.properties.roleId=<role id>
    ```
    !!! note
        In production, you should always use the vault address with TLS enabled.

> 5. Add the following lines to the `<APIM_HOME>/repository/conf/log4j2.properties` file:

Reviewer: ```suggestion
5. Add the following lines to the `<API-M_HOME>/repository/conf/log4j2.properties` file:
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    ```properties
    logger.org-wso2-carbon-securevault-hashicorp.name=org.wso2.carbon.securevault.hashicorp
    logger.org-wso2-carbon-securevault-hashicorp.level=INFO
    logger.org-wso2-carbon-securevault-hashicorp.additivity=false
    logger.org-wso2-carbon-securevault-hashicorp.appenderRef.CARBON_CONSOLE.ref = CARBON_CONSOLE
    ```

>     Then append `org-wso2-carbon-securevault-hashicorp` to the `loggers` list in the same file as follows:

Reviewer: ```suggestion
6. Then append `org-wso2-carbon-securevault-hashicorp` to the `loggers` list in the same file as follows.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

>         ```properties

Reviewer: The alignment needs to be adjusted.

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

        loggers = AUDIT_LOG, trace-messages, ..., org-wso2-carbon-securevault-hashicorp
        ```

> ### Step 3: Update passwords with their aliases

Reviewer: ```suggestion
### Step 3 - Update passwords with their aliases
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Open the `deployment.toml` file in the `<APIM_HOME>/repository/conf/` directory and add

Reviewer: ```suggestion
1. Open the `deployment.toml` file in the `<API-M_HOME>/repository/conf/` directory and add
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

   the `[secrets]` configuration section **at the bottom of the file** as shown below.
   Give an alias for the passwords and put the value as blank (`""`).

    ```toml
    [secrets]
    admin_password = ""
    ```

2. Add the encrypted password alias to the relevant sections in the `deployment.toml`
   file by using a placeholder: `$secret{alias}`. For example:

    ```toml
    [super_admin]
    username="admin"
    password="$secret{admin_password}"
    ```

> ### Step 4: Start the server

Reviewer: ```suggestion
### Step 4 - Start the server
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

> 1. Provide the `Secret ID` to the prompted message in the console or by creating a new file in the `<APIM_HOME>` 

Reviewer: ```suggestion
1. Provide the `Secret ID` to the prompted message in the console or by creating a new file in the `<API-M_HOME>` 
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

   directory. The file should be named according to your Operating System.

    === "Linux"
        ```

>         The file name should be `hashicorpSecretId-tmp`.

Reviewer: Let's change the formatting as suitable for a statement.

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

        ```
      
    === "Windows"
        ```
        The file name should be `hashicorpSecretId-tmp.txt`.
        ```
        
!!! note

>     When you add "tmp" to the file name, note that this will automatically get deleted from the file system after the server starts. Alternatively, if you want to retain the password file after the server starts, the file should be named as follows:

Reviewer: ```suggestion
    When you add `tmp` to the file name, note that this will automatically get deleted from the file system after the server starts. Alternatively, if you want to retain the password file after the server starts, the file should be named as follows:
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

    === "Linux"
        ```

>         The file name should be `hashicorpSecretId-persist`.

Reviewer: Let's change the formatting as suitable for a statement.

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a

        ```
      
    === "Windows"
        ```
        The file name should be `hashicorpSecretId-persist.txt`.
        ```

> 2. Start the WSO2 Identity Server and enter the keystore password at startup when prompted:

Reviewer: ```suggestion
2. Start the WSO2 Identity Server and enter the keystore password at startup when prompted.
```

Author: corrected in 1d8bdffa01d44e5da79703dd2060f0a621bb659a
